spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+libsl "1.1.0";
+
+library "std:collections" language "Java" version "11" url "-";
+
+import "java-common.lsl";
+import "list-actions.lsl";
+import "java/util/interfaces.lsl"
+import "java/util/function/interfaces.lsl"
+
+//For generation this imports are needed:
+include java.util.AbstractList;
+include java.util.ArrayList;
+include java.util.Collection;
+include java.util.Iterator;
+include java.util.List;
+include java.util.ListIterator;
+include java.util.NoSuchElementException;
+include java.util.Objects;
+include java.util.RandomAccess;
+include java.util.function.Consumer;
+include java.util.function.Predicate;
+include java.util.function.UnaryOperator;
+include java.util.stream.Stream;
+
+// automata
+
+@extends("java.util.AbstractList")
+@implements(["java.util.List", "java.util.RandomAccess","java.lang.Cloneable","java.io.Serializable"])
+@WrapperMeta(
+    src="java.util.ArrayList",
+    dst="ru.spbpu.libsl.overrides.collections.ArrayList",
+    forceMatchInterfaces=true)
+automaton ArrayList: int (
+    @private @static @final var serialVersionUID: long = 8683452581122892189,
+    @transient var storage: List<Object> = null,
```

Comment:
@dog-m But why this is "transient" in the original class ? 
`transient Object[] elementData;`

[Why does ArrayList use transient storage?](https://stackoverflow.com/questions/9848129/why-does-arraylist-use-transient-storage)

As I understood we have fixed size of list (storage.size() == length == true), that's why we can make it NOT transient... Not like in the original ArrayList class. This is right ?

Reply:
If we imagine the concept of `list` in relation to not-so-long-ago agreed piece of *target API*, then it does have the `size: () -> int` method. But for every spec we have created to this moment we used additional *variable* to store the size of the list. And, yes, at some point we did have an `action LIST_SIZE` or something similar. But calling every time such expensive method (involving unknown amount of some SMT interfacing stuff) is not a great idea.

In short, each list object have some internal knowledge about its size, so no need in storing additional `length`s or `size`es.

And, no, the approach from the original ArrayList does not apply to our specification because we use dynamic lists instead of fixed arrays. So no "excess empty tail" problem.

In general, I have no clear answer on how to do serialization yet because we haven't deeply researched the related machinery yet (this includes *special private methods*).

I insist on making this field/variable non-transient just in case we decide to use some automatic serialization mechanism.